src/backend/Buffer.cpp

```diff
@@ -155,58 +148,31 @@ namespace backend {
         mMapUserdata = 0;
     }
 
-    bool BufferBase::IsFrozen() const {
-        return mIsFrozen;
-    }
-
-    bool BufferBase::HasFrozenUsage(nxt::BufferUsageBit usage) const {
-        return mIsFrozen && (usage & mAllowedUsage);
-    }
-
-    bool BufferBase::IsUsagePossible(nxt::BufferUsageBit allowedUsage, nxt::BufferUsageBit usage) {
-        bool allowed = (usage & allowedUsage) == usage;
-        bool readOnly = (usage & kReadOnlyBufferUsages) == usage;
-        bool singleUse = nxt::HasZeroOrOneBits(usage);
-        return allowed && (readOnly || singleUse);
-    }
-
-    bool BufferBase::IsTransitionPossible(nxt::BufferUsageBit usage) const {
-        if (mIsFrozen || mIsMapped) {
+    bool BufferBase::ValidateMapBase(uint32_t start,
+                                     uint32_t size,
+                                     nxt::BufferUsageBit requiredUsage) {
+        if (start + size > GetSize()) {
+            mDevice->HandleError("Buffer map read out of range");
             return false;
         }
-        return IsUsagePossible(mAllowedUsage, usage);
-    }
 
-    void BufferBase::UpdateUsageInternal(nxt::BufferUsageBit usage) {
-        ASSERT(IsTransitionPossible(usage));
-        mCurrentUsage = usage;
-    }
-
-    void BufferBase::TransitionUsage(nxt::BufferUsageBit usage) {
-        if (!IsTransitionPossible(usage)) {
-            mDevice->HandleError("Buffer frozen or usage not allowed");
-            return;
+        if (mIsMapped) {
+            mDevice->HandleError("Buffer already mapped");
+            return false;
         }
-        TransitionUsageImpl(mCurrentUsage, usage);
-        mCurrentUsage = usage;
-    }
 
-    void BufferBase::FreezeUsage(nxt::BufferUsageBit usage) {
-        if (!IsTransitionPossible(usage)) {
-            mDevice->HandleError("Buffer frozen or usage not allowed");
-            return;
+        if (!(mAllowedUsage & requiredUsage)) {
+            mDevice->HandleError("Buffer needs the correct map usage bit");
+            return false;
         }
-        mAllowedUsage = usage;
-        TransitionUsageImpl(mCurrentUsage, usage);
-        mCurrentUsage = usage;
-        mIsFrozen = true;
+
+        return true;
     }
 
     // BufferBuilder
 
     enum BufferSetProperties {
         BUFFER_PROPERTY_ALLOWED_USAGE = 0x1,
-        BUFFER_PROPERTY_INITIAL_USAGE = 0x2,
         BUFFER_PROPERTY_SIZE = 0x4,
```

Comment:
Any reason to keep this as 0x4 instead of 0x2?

Reply:
Done